src/components/vm/disks/diskAdd.jsx

```diff
@@ -583,16 +585,35 @@ export class AddDiskModalBody extends React.Component {
             storagePool = storagePools.find(pool => pool.name === this.state.storagePoolName);
             volume = storagePool.volumes.find(vol => vol.name === this.state.existingVolumeName);
             isVolumeUsed = getStorageVolumesUsage(vms, storagePool);
+        } else if (this.state.mode === CUSTOM_PATH) {
+            optionallyResolveFormat = cockpit.spawn(["head", "--bytes", "3", this.state.file], { binary: true, err: "message" })
+                    .then(output => {
+                        let magic = "";
+                        output.map(m => { magic += String.fromCharCode(m) });
+                        switch (magic) {
+                        case "QFI":
+                            format = "qcow2";
+                            break;
+                        default:
+                            format = "raw";
+                            break;
+                        }
+                    })
+                    .catch(e => {
+                        console.warn("could not execute qemu-img", e.toString());
+                        // fallback to raw.
+                        format = "raw";
+                    });
```

Comment:
From the conversation we had at #virt IRC, I believed it was recommended that we should in the end always allow for users to specify a format, and use this autodetection. So in practice, this autodetection could be use to change the default value of "Format" in the UI, but in the end allow the user to change it, no? As there are other security issues (e.g. you can put qcow2 header over raw wile aswell)
![Screenshot from 2022-03-31 18-02-26](https://user-images.githubusercontent.com/42733240/161099507-c18cfd80-45a5-44d0-af9b-c647ccb8e5dd.png)

Reply:
I'd like to circle back on this. How exactly is a manual format specification better? How would the user decide what format it is? They can decide either by file extension (which is not reliable for malicious images either), or by running `qemu-img info` (which we got told we shouldn't do). So looking at the magic seems like the safest way. We can certainly *show* the format somewhere, so that at least there's some raised eyebrow if a user specifies `somedisk.raw` and the format is actually "qcow" or "vmdk".

The potential security issue is not about the format autodetection, but about a malicious qcow image with an arbitrary backing image. That's what we should detect, but that requires `qemu-img info`. And frankly, if you can run arbitrary code on the host with that, that's a bug in qemu-img. That should just dissect the format, not actually run anything.